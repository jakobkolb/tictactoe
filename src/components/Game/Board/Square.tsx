import React from 'react'
import { PLAYERS } from '../../../types'
import * as R from 'ramda'

interface SquareProps {
  player: PLAYERS | null
  onClick: () => void
}

const playersToChars = {
  [PLAYERS.X]: 'X',
  [PLAYERS.O]: 'O',
}

export const symbol: (player: PLAYERS | null) => string | null = R.pipe<
  PLAYERS | null,
  string | undefined,
  string | null
>(R.flip(R.prop)(playersToChars) as any, R.defaultTo(null))

const playersToColors = {
  [PLAYERS.X]: '#178fc2',
  [PLAYERS.O]: '#912c1a',
}

export const color: (player: PLAYERS | null) => string = R.pipe<
  PLAYERS | null,
  string | undefined,
  string
>(R.flip(R.prop)(playersToColors) as any, R.defaultTo('#ffffff'))

export const Square: React.SFC<SquareProps> = ({ onClick, player }) => (
  <button className="square" onClick={onClick} style={{ color: color(player) }}>
    {symbol(player)}
  </button>
)
