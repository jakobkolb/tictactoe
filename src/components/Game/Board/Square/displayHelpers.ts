import { PLAYERS } from '../../../../types'
import * as R from 'ramda'

export const playersToChars = {
  [PLAYERS.X]: 'X',
  [PLAYERS.O]: 'O',
}
export const symbol: (player: PLAYERS | null) => string | null = R.pipe<
  PLAYERS | null,
  string | undefined,
  string | null
>(R.flip(R.prop)(playersToChars) as any, R.defaultTo(null))
const playersToColors = {
  [PLAYERS.X]: '#178fc2',
  [PLAYERS.O]: '#912c1a',
}
export const color: (player: PLAYERS | null) => string = R.pipe<
  PLAYERS | null,
  string | undefined,
  string
>(R.flip(R.prop)(playersToColors) as any, R.defaultTo('#ffffff'))
