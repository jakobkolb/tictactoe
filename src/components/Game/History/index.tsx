import React from 'react'

import { JumpToMoveButtons } from './JumpToMoveButtons'
import { JumpToStartButton } from './JumpToStartButton'

export const HistoryNavigation: React.SFC = () => (
  <ol>
    <JumpToStartButton />
    <JumpToMoveButtons />
  </ol>
)
