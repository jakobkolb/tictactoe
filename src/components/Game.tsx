import React from 'react'
import '../index.css'
import Board from './board'

const COLORS = { xColor: '#178fc2', oColor: '#912c1a', neutralColor: '#ffffff' }
const SYMBOLS = { X: 'X', O: 'O' }

interface GameProps {
  value?: number
  name?: string
}

interface GameState {
  history: {
    squares: (string | null)[]
    colors: string[]
    xIsNext: boolean
  }[]
  stepNumber: number
}

export const calculateWinner = (squares: (string | null)[]): string | null => {
  const lines = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6],
  ]
  for (let i = 0; i < lines.length; i++) {
    const [a, b, c] = lines[i]
    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
      return squares[a]
    }
  }
  return null
}

class Game extends React.Component<GameProps, GameState> {
  constructor(props: GameProps) {
    super(props)
    this.state = {
      history: [
        {
          squares: Array(9).fill(null),
          colors: Array(9).fill(COLORS.neutralColor),
          xIsNext: true,
        },
      ],
      stepNumber: 0,
    }
  }
  handleClick(i: number): void {
    const stepNumber = this.state.stepNumber
    const history = this.state.history.slice(0, stepNumber + 1)
    const current = history[history.length - 1]
    const xIsNext = current.xIsNext
    const squares = current.squares.slice()
    if (calculateWinner(squares) || squares[i]) {
      return
    }
    squares[i] = xIsNext ? SYMBOLS.X : SYMBOLS.O
    const colors = current.colors.slice()
    colors[i] = xIsNext ? COLORS.xColor : COLORS.oColor
    this.setState({
      history: history.concat([
        {
          squares: squares,
          colors: colors,
          xIsNext: !xIsNext,
        },
      ]),
      stepNumber: history.length,
    })
  }

  jumpTo(step: any) {
    this.setState({
      stepNumber: step,
    })
  }

  render() {
    const history = this.state.history
    const stepNumber = this.state.stepNumber
    const current = history[stepNumber]
    const xIsNext = current.xIsNext
    const winner = calculateWinner(current.squares)

    const moves = history.map((step, move) => {
      const desc = move ? `Go to move # ${move}` : `Go to game start`
      return (
        <li key={move}>
          <button onClick={() => this.jumpTo(move)}>{desc}</button>
        </li>
      )
    })

    let status
    if (winner) {
      status = `Winner: ${winner}`
    } else {
      status = `Next player: ${xIsNext ? SYMBOLS.X : SYMBOLS.O}`
    }
    return (
      <div className="game">
        <div className="game-board">
          <Board
            squares={current.squares}
            colors={current.colors}
            onClick={(i) => this.handleClick(i)}
          />
        </div>

        <div className="game-info">
          <div>{status}</div>
          <ol>{moves}</ol>
        </div>
      </div>
    )
  }
}

export default Game
