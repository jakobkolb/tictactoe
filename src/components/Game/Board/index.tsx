import React from 'react'
import { Square } from './Square'
import { PLAYERS } from 'types'

interface BoardProps {
  squares: (PLAYERS | null)[]
  onClick: (i: number) => void
}

export const Board: React.SFC<BoardProps> = (props: BoardProps) => {
  const renderSquare = (i: number) => {
    return (
      <Square
        player={props.squares[i]}
        onClick={(): void => props.onClick(i)}
      />
    )
  }

  return (
    <div>
      <div className="board-row">
        {renderSquare(0)}
        {renderSquare(1)}
        {renderSquare(2)}
      </div>
      <div className="board-row">
        {renderSquare(3)}
        {renderSquare(4)}
        {renderSquare(5)}
      </div>
      <div className="board-row">
        {renderSquare(6)}
        {renderSquare(7)}
        {renderSquare(8)}
      </div>
    </div>
  )
}
