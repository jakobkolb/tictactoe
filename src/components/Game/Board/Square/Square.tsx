import React from 'react'
import { Dispatch } from 'redux'
import { Board, GameState, PLAYERS } from 'types'
import * as R from 'ramda'
import { getCurrentBoardFromState } from 'helpers/selectors'
import * as actions from 'helpers/actions'
import { connect } from 'react-redux'
import { color, symbol } from './displayHelpers'

interface OwnProps {
  location: number
}

interface StateProps {
  player: PLAYERS | null
}

interface DispatchProps {
  onClick: () => void
}

interface SquareProps extends OwnProps, StateProps, DispatchProps {}

const SquareComponent: React.SFC<SquareProps> = ({ onClick, player }) => (
  <button className="square" onClick={onClick} style={{ color: color(player) }}>
    {symbol(player)}
  </button>
)

const mapStateToProps = (state: GameState, { location }: OwnProps) => {
  const player = R.pipe<GameState, Board, PLAYERS | null>(
    getCurrentBoardFromState,
    R.prop((location as unknown) as string) as (b: Board) => PLAYERS | null,
  )(state)
  return { player: player }
}

const mapDispatchToProps = (
  dispatch: Dispatch,
  { location }: OwnProps,
): DispatchProps => ({
  onClick: (): void => dispatch(actions.createClickAction(location)),
})

export const Square = connect(
  mapStateToProps,
  mapDispatchToProps,
)(SquareComponent)
