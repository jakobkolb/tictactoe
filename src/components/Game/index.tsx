import React from 'react'
import { GameBoard } from './Board'
import * as R from 'ramda'
import { GameState, Board, History } from 'types'

import { getStatus } from 'helpers'
import * as selectors from 'helpers/selectors'
import * as actions from 'helpers/actions'
import { connect } from 'react-redux'
import { HistoryNavigation } from './History'

const mapDispatchToProps = {
  makeMove: actions.createClickAction,
}

const mapStateToProps: (state: GameState) => StateProps = R.applySpec({
  squares: selectors.getCurrentBoardFromState,
  history: selectors.getHistory,
  status: getStatus,
})

interface StateProps {
  history: History
  status: string
  squares: Board
}

interface DispatchProps {
  makeMove: (boardIndex: number) => void
}

interface GameComponentProps extends StateProps, DispatchProps {}

const GameComponent: React.SFC<GameComponentProps> = ({
  history,
  status,
  makeMove,
  squares,
}) => (
  <div className="game">
    <div className="game-board">
      <GameBoard squares={squares} onClick={makeMove} />
    </div>

    <div className="game-info">
      <Status status={status} />
      <HistoryNavigation history={history} />
    </div>
  </div>
)

const Status: React.SFC<{ status: string }> = ({ status }) => (
  <div>{status}</div>
)

export const Game: React.SFC = connect(
  mapStateToProps,
  mapDispatchToProps,
)(GameComponent)
