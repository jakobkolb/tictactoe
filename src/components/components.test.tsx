import React from 'react'
import { calculateWinner } from './Game'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Square from './square'
import Board from './board'

configure({ adapter: new Adapter() })

describe('calculateWinner', () => {
  test('should return winner if she has three in line', () => {
    expect(
      calculateWinner(['X', 'X', 'X', null, 'O', 'O', null, null, null, null]),
    ).toEqual('X')
  })
  test('should return null, if noone has three in line', () => {
    calculateWinner(['X', 'X', 'O', null, 'O', 'O', null, 'X', null, null])
  })
})

describe('Square', () => {
  test('should render correctly', () => {
    const wrapper = shallow(<Square value={'X'} onClick={(): void => void 0} />)
    expect(wrapper).toMatchSnapshot()
  })
  test('should contain symbol, if passed one', () => {
    const wrapper = shallow(<Square value={'o'} onClick={(): void => void 0} />)
    expect(wrapper.text()).toMatch('o')
  })
  test('should call callback method when clicked', () => {
    const onClick = jest.fn()
    const wrapper = shallow(<Square value={null} onClick={onClick} />)
    wrapper.simulate('click')
    expect(onClick).toHaveBeenCalled()
  })
})

describe('Board', () => {
  const squares = ['X', 'X', 'X', null, 'O', 'O', null, null, null, null]
  const colors = [
    '#178fc2',
    '#178fc2',
    '#ffffff',
    '#ffffff',
    '#912c1a',
    '#912c1a',
    '#ffffff',
    '#ffffff',
    '#ffffff',
  ]
  test('should render correctly', () => {
    const onClick = jest.fn()
    const wrapper = shallow(
      <Board squares={squares} colors={colors} onClick={onClick} />,
    )
    expect(wrapper).toMatchSnapshot()
  })
  test('should call callback with square id when clicked', () => {
    const onClick = jest.fn()
    const wrapper = shallow(
      <Board squares={squares} colors={colors} onClick={onClick} />,
    )
    wrapper.find(Square).at(3).simulate('click')
    expect(onClick).toHaveBeenCalledWith(3)
  })
})
