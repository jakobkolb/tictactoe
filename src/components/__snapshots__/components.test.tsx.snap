// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should render correctly 1`] = `ShallowWrapper {}`;

exports[`Square should render correctly 1`] = `ShallowWrapper {}`;
